ticdc(codec): add avro refactor design (#5413)
pingcap/tiflow: docs/design/2022-05-12-ticdc-avro-protocol-refactor.md

## Motivation or Background

Apache Avro™ is a data serialization system with rich data structures and a compact binary data format. Avro relies on schemas, schemas is managed by schema-registry. Avro is a common data format in streaming systems, supported by Confluent, Flink, Debezium, etc.

## Detailed Design

### New Config Items

| Config item                        | Option values          | Default | Explain                                                                                                                                                                                                                                                                                                         |
| ---------------------------------- | ---------------------- | ------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | protocol                           | canal-json / flat-avro | -       | Specify the message format which output to the kafka.<br>The `flat-avro` option means that using the avro format design by this document.                                                                                                                                                                       |

Reviewer: ```suggestion
| protocol                           | canal-json / flat-avro | -       | Specify the format of messages written to Kafka.<br>The `flat-avro` option means that using the avro format design by this document.                                                                                                                                                                       |
```

| enable-tidb-extension              | true / false           | false   | Append tidb extension fields into avro message or not.                                                                                                                                                                                                                                                          |
| schema-registry                    | -                      | -       | Specifies the schema registry endpoint.                                                                                                                                                                                                                                                                         |
| avro-decimal-handling-mode         | precise / string       | precise | Specifies how the TiCDC should handle values for DECIMAL columns:<br>`precise` option represents encoding decimals as precise bytes.<br>`string` option encodes values as formatted strings, which is easy to consume but semantic information about the real type is lost.                                     |

> | avro-bigint-unsigned-handling-mode | long / string          | long    | Specifies how the TiCDC should handle values for UNSIGNED BIGINT columns:<br>`long` represents values by using Avro long(64-bit signed integer) which might overflow but which is easy to use in consumers.<br>`string` represents values by string which is precision but which is need to parse by consumers. |

Reviewer: ```suggestion
| avro-bigint-unsigned-handling-mode | long / string          | long    | Specifies how the TiCDC should handle values for UNSIGNED BIGINT columns:<br>`long` represents values by using Avro long(64-bit signed integer) which might overflow but which is easy to use in consumers.<br>`string` represents values by string which is precise but which needs to be parsed by consumers. |
```

#### Key Schema

```
{
    "name":"{{RecordName}}",
    "namespace":"{{Namespace}}",
    "type":"record",
    "fields":[
        {{ColumnValueBlock}},
        {{ColumnValueBlock}},
    ]
}
```

- `{{RecordName}}` represents full qualified table name.
- `{{ColumnValueBlock}}` represents a JSON block, which defines a column value of a key.
- Key only includes the valid index fields.

#### Value Schema

```
{
    "name":"{{RecordName}}",
    "namespace":"{{Namespace}}",
    "type":"record",
    "fields":[
        {{ColumnValueBlock}},
        {{ColumnValueBlock}},
        {
            "name":"_tidb_op",
            "type":"string"
        },
        {
            "name":"_tidb_commit_ts",
            "type":"long"
        },
        {
            "name":"_tidb_commit_physical_time",
            "type":"long"
        }
    ]
}
```

- `{{RecordName}}` represents full qualified table name.
- `{{ColumnValueBlock}}` represents a JSON block, which defines a column value of a key.

> - `_tidb_op` used to distinguish between INSERT or UPDATE events, optional values are "c" / "u".

Reviewer: ```suggestion
- `_tidb_op` is used to distinguish between INSERT or UPDATE events, optional values are "c" / "u".
```

- avro/flat-avro protocol
- avro/flat-avro protocol & true/false/invalid enable-tidb-extension
- avro/flat-avro protocol & precise/string/invalid avro-decimal-handling-mode
- avro/flat-avro protocol & long/string/invalid avro-bigint-unsigned-handling-mode
- avro/flat-avro protocol & valid/invalid schema-registry

#### Data Mapping Tests

- With protocol=avro&enable-tidb-extension=false&avro-decimal-handling-mode=precise&avro-bigint-unsigned-handling-mode=long, all generated schema and data are correct.
- With enable-tidb-extension=true, schema and value will have \_tidb_op, \_tidb_commit_ts, \_tidb_commit_physical_time fields.
- With avro-decimal-handling-mode=string，decimal field generates string schema and data.
- With avro-bigint-unsigned-handling-mode=string, bigint unsigned generates string schema and data.

#### DML Tests

- Insert row and check the row in downstream database.
- Update row and check the row in downstream database.
- Delete row and check the row in downstream database.

#### Schema Tests

- When the schema is not in schema registry, a fresh new schema is created.
- When the schema is in schema registry and pass compatibility check, a new version is created.

> - When the schema is in schema registry and not pass compatibility check, reports error.

Reviewer: ### 
```suggestion
- When the schema is in schema registry and cannot pass compatibility check, reports error.
```
